resource-aggregate: add Delete call to eventstore

Danielius1922 opened the pull request:
Add Delete call to eventstore which deletes all documents with given groupIds.

Danielius1922 committed:
resource-aggregate: add Delete call to eventstore

Add Delete call to eventstore which deletes all documents
with given groupIds.

jkralik commented on resource-aggregate/cqrs/eventstore/mongodb/delete.go:
```diff
@@ -0,0 +1,59 @@
+			groupIDKey: deviceId,
+		})
+		if err != nil {
+			errors = append(errors, fmt.Errorf("failed to delete documents with groupid(%v): %v", deviceId, err))
```
pls return the error - it happens when something bad happens.

Danielius1922 replied:
fixed

Danielius1922 committed:
resource-aggregate: refactor device access validation

Danielius1922 committed:
resource-aggregate: add DeleteDevices call

Danielius1922 committed:
authorization: change DeleteDevice to DeleteDevices

Rework Delete call to enable deletion of multiple deviceIds
at the same time.

Danielius1922 committed:
resource-aggregate: rework Delete call

Previously the implementation ran delete one-by-one
by deviceID and returned a list of device ids that had
at least one document deleted. However, this is not a
useful output because the caller cannot use it to determine
whether all documents with the deviceID were deleted,
or whether documents with given ID were not deleted because
no such documents existed before the call.
Instead of the output from the Delete command the db should
be queried again to get the current state.

Danielius1922 committed:
grpc-gateway: add DeleteDevices call

Danielius1922 committed:
resource-aggregate: update DeleteDevices

Sending DeleteDevicesRequest with empty DeviceIds list
is interpreted as requesting to delete all owned devices

Danielius1922 committed:
grpc-gateway: update DeleteDevices call

Delete all owned devices if empty deviceIdFilter is
used.

jkralik commented on grpc-gateway/config.yaml:
```diff
@@ -35,6 +35,19 @@ apis:
           useSystemCAPool: false
 clients:
+  authorizationServer:
+    ownerClaim: sub
```
do you need ownerClaim ?

Danielius1922 replied:
ownerClaim removed

jkralik commented on authorization/service/deleteDevices.go:
```diff
@@ -0,0 +1,70 @@
+	// TODO validate jwt token -> only jwt token is supported
+
+	var deletedDeviceIds []string
+	for _, deviceId := range request.DeviceIds {
```
if deviceIDs are empty you don't proceed any deletein auth service. but in ra you delete all resources of user.

jkralik commented on resource-aggregate/service/deleteDevices.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, log.LogAndReturnError(kitNetGrpc.ForwardErrorf(codes.InvalidArgument, "cannot delete devices: invalid request"))
+	}
+
+	owner, ownedDevices, err := r.getOwnedDevices(ctx, deviceIds)
```
if deviceIds is not set, you delete all devices at RA but in auth, you don't delete any if the request doesn't contain deviceIds

jkralik reviewed (changes requested):
pls fix inconsistency of delete between RA and AUTH service for empty deviceIds.

Danielius1922 committed:
http-gateway: add handling of DELETE command for /api/v1/devices

Danielius1922 committed:
resource-aggregate: UML for DeleteDevices command

jkralik commented on grpc-gateway/service/deleteDevices.go:
```diff
@@ -0,0 +1,81 @@
+		return &pb.DeleteDevicesResponse{
+			DeviceIds: deleted,
+		}, nil
+	}
```
if `deleteAllOwned==true` and `len(deleted) == 0` then we want to call `r.authorizationClient.DeleteDevices(ctx, &cmdAS)`

jkralik replied:
we always want to delete all devices from authorizationClient with same argument as in `cmdRA.DeviceIds`

Danielius1922 committed:
DeleteDevices: code review fixes

jkralik commented on coap-gateway/service/signOff.go:
```diff
@@ -118,14 +118,18 @@ func signOffHandler(req *mux.Message, client *Client) {
-		DeviceId: signOffData.deviceID,
-		UserId:   signOffData.userID,
-	}); err != nil {
+	resp, err := client.server.asClient.DeleteDevices(ctx, &pb.DeleteDevicesRequest{
```
maybe we need to call deleteDevices to RA

Danielius1922 replied:
Yes, I think it makes sense. Fixed.

Danielius1922 committed:
Refactor: move Intersection function to pkg

Danielius1922 committed:
pkg: add Split and Unique function to strings package

Danielius1922 committed:
grpc-gateway: update DeleteDevices

Sent the same input to ResourceAggregate and Authorization
service from DeleteDevices in grpc-gw

Danielius1922 committed:
coap-gateway: clean-up eventstore on device signOff

Remove all documents for given device from eventstore
when it signs off.

ondrejtomcik commented on authorization/client/userDevices.go:
```diff
@@ -169,6 +169,19 @@ func (d *UserDevicesManager) UpdateUserDevices(ctx context.Context, userID strin
 	return userDevices[userID], nil
 }
 
+func (d *UserDevicesManager) ChecksUserDevices(userID string, deviceIDs []string) bool {
```
Just came to my mind, should we really name it User? Isn't it the owner? Also RPC calls, AddUserDevices. It's assigning to the owner, and the user doesn't have to necessarily be an end-user, or? @Danielius1922 , @jkralik what do you think?

Danielius1922 replied:
`Owner` seems better, I think @jkralik suggested something similar: renaming GetUserDevices to GetOwnerDevices. If it better expresses what it does I think we should rename it, though we should rename it in all places and not just in one or two.

ondrejtomcik replied:
Discussed, we'll do it but in separate MR.

Danielius1922 committed:
Fix memory issues and leaks

ondrejtomcik commented on authorization/service/addDevice.go:
```diff
@@ -54,11 +54,12 @@ func (s *Service) AddDevice(ctx context.Context, request *pb.AddDeviceRequest) (
-		uid, err := grpc.ParseOwnerFromJwtToken(s.ownerClaim, token)
-		if err == nil && uid != serviceOwner {
-			owner = uid
+	if owner == "" {
```
This will be refactored later? As we discussed owner will be taken only from the token.

Danielius1922 replied:
Yes, it must be done later, because if I remove this it will break tests. We need to develop a mechanism with inserts a custom userId into a token and update the tests, not sure if there is a story for this in monday

ondrejtomcik commented on authorization/service/deleteDevices.go:
```diff
@@ -0,0 +1,135 @@
+				Owner:     owner,
+				DeviceIds: deviceIDs,
+				AuditContext: &events.AuditContext{
+					UserId: owner,
```
This should be always the "sub" claim. Config.ownerClaim can be changed from "sub" to something else and in such a case, we lose the information on who did this change.

Danielius1922 replied:
Shouldn't it be removed from configuration then if it shouldn't be changed?

ondrejtomcik replied:
No, because the `event.Owner` should match `ownerClaim` claim value. `AuditContext` always "sub". Reason:
`ownerClaim` is set to tenantId - organization identifier. That means that the device is owned by the organization and any employee of the organization can retrieve it. If you pass the same id (tenantId) to the AuditContext, all you can track back is that this organization deleted it. But you don't know which bad guy did this mistake, because you didn't take his userId, which is always in "sub" claim :) So it's a good feature for bad guys, but not for bad security managers :D

Danielius1922 replied:
In the case that ownerClaim != 'sub' where will both claims be. In the JTW? Is it guaranteed to always have both non-empty? Or what should I do when 'sub' or custom claim is empty?

ondrejtomcik replied:
All gateways have to verify 
if ownerClaim == "sub"
  jwt.sub must be present
if ownerClaim == "abc"
  jwt.sub && jwt.abc must be present

ondrejtomcik replied:
Will be covered in the separate task. @Danielius1922 please add todo at this line

Danielius1922 replied:
TODO added

ondrejtomcik commented on authorization/service/deleteDevices.go:
```diff
@@ -0,0 +1,135 @@
+	defer tx.Close()
+
+	owner := request.UserId
+	if owner == "" {
```
Same applies here.

ondrejtomcik replied:
todo comment

Danielius1922 replied:
TODO added

ondrejtomcik commented on authorization/service/deleteDevices.go:
```diff
@@ -0,0 +1,135 @@
+		return nil, log.LogAndReturnError(status.Errorf(codes.InvalidArgument, "cannot delete devices: invalid UserId"))
+	}
+
+	// TODO validate jwt token -> only jwt token is supported
```
Is this in monday? Which task?

Danielius1922 replied:
Hmm copy & paste from original implementation in authorization/service/deleteDevice.go, not sure if it is relevant anymore...?

ondrejtomcik replied:
Validation of JWT is done somewhere else, right? Then you can delete it :)

ondrejtomcik commented on grpc-gateway/config.yaml:
```diff
@@ -35,6 +35,18 @@ apis:
           certFile: "/secrets/public/cert.crt"
           useSystemCAPool: false
 clients:
+  authorizationServer:
```
Please update also deployment config which is still part of this repo.

ondrejtomcik replied:
@rafajpet FYI

Danielius1922 replied:
Could you please link it? I don't know what you mean.

ondrejtomcik replied:
https://github.com/plgd-dev/cloud/blob/v2/docs/guide/deployment/grpc-gateway.md

Danielius1922 replied:
grpc-gw documentation updated

ondrejtomcik commented on http-gateway/swagger.yaml:
```diff
@@ -49,6 +49,42 @@ paths:
+    delete:
+      tags:
+        - 'Devices'
+      summary: 'Delete owned devices'
```
you can skip word owned. It might raise the question "what is owned device", "if there can be connected not owned device", and so on. Each device connected to the plgd cloud is owned.

ondrejtomcik commented on resource-aggregate/.plantuml:
```diff
@@ -296,4 +296,28 @@ ResourceAggregate --> Bus : DeviceMetadataUpdated (Canceled=true)
+activate Server
+activate Gateway
+activate AuthServer
+Server -> Gateway : DeleteDevicesRequest
```
Is it really a device sending the delete request? The device sends SignOff. This example should be about the client, nope?

ondrejtomcik commented on resource-aggregate/.plantuml:
```diff
@@ -296,4 +296,28 @@ ResourceAggregate --> Bus : DeviceMetadataUpdated (Canceled=true)
+participant "Authorization Server" as AuthServer
+
+activate ResourceAggregate
+activate Server
```
activation represents processing of the request. You have to do it step by step. :)

ondrejtomcik commented on resource-aggregate/.plantuml:
```diff
@@ -296,4 +296,28 @@ ResourceAggregate --> Bus : DeviceMetadataUpdated (Canceled=true)
+activate Gateway
+activate AuthServer
+Server -> Gateway : DeleteDevicesRequest
+Gateway-> ResourceAggregate : DeleteDevicesRequest
```
Could you please include also EventBus into this diagram and visualize events which these components, RA, AS are sending?

Danielius1922 replied:
Not sure what you mean here. This actually deletes all events for given devices from the RA database and all devices from the Auth database ...?

ondrejtomcik commented on resource-aggregate/.plantuml:
```diff
@@ -296,4 +296,28 @@ ResourceAggregate --> Bus : DeviceMetadataUpdated (Canceled=true)
+ResourceAggregate -> Gateway : DeleteDevicesResponse
+Gateway -> AuthServer : DeleteDevicesRequest
+AuthServer -> Gateway : DeleteDevicesResponse
+Gateway -> Server : DeleteDevicesResponse
```
Could you please include this diagram also in the plgd.dev docs, `archtiecture/component-overview/CoAP Gateway` with header Device removed by the user from the plgd Cloud ?

And make it more detailed, including interaction with the device after delete:
example:
```
...
CGW -> Server: Disconnect
destroy Server
deactivate CGW
Server -> CGW: SignUp
activate Server
activate CGW
CGW -> AS: ??? question on chat
...
CGW -> Server: Unauthorized
CGW -> Server: Disconnect
destroy Server
deactivate CGW
Server -> Server: Remove cloud credentials
```

ondrejtomcik replied:
This MR can be closed ofc, this comment is not blocking this one :)

ondrejtomcik commented on authorization/service/deleteDevices.go:
```diff
@@ -0,0 +1,135 @@
+		deletedDeviceIds = append(deletedDeviceIds, deviceId)
+	}
+
+	if err := s.publishDevicesUnregistered(ctx, owner, deletedDeviceIds); err != nil {
```
I am not really sure if we can send this event. At the moment we introduce sharing of devices, or customers will implement it in their own auth service, it's impossible to distinguish if the device was really deleted or if sharing for user B was stopped, but it's still in the system for user A. Plan was to send this event if device sharing is stopped for user B by user A for example. But then, coap gateway listening on these events cannot decide if it should disconnect it or not. Does this explanation make sense?
@Danielius1922 @jkralik

Danielius1922 committed:
Minor code review updates

ondrejtomcik committed:
update delete devices plantuml

ondrejtomcik committed:
update commands docs

Danielius1922 committed:
grpc-gateway: update configuration documentation
